test/functional/readconcern_tests.js

```diff
@@ -1,498 +1,361 @@
-      client.connect(function(err, client) {
-        expect(err).to.not.exist;
-
-        var db = client.db(configuration.db);
-        test.deepEqual({ level: 'majority' }, db.s.readConcern);
-
-        // Get a collection
-        var collection = db.collection('readConcernCollection');
-        // Validate readConcern
-        test.deepEqual({ level: 'majority' }, collection.s.readConcern);
-        // Perform a find using the readConcern
-        listener.on('started', function(event) {
-          if (event.commandName === 'find') started.push(event);
-        });
-
-        // Execute find
-        collection.find().toArray(function(err) {
-          test.equal(null, err);
-          test.equal(1, started.length);
-          test.deepEqual({ level: 'majority' }, started[0].command.readConcern);
-
-          listener.uninstrument();
-          client.close();
-          done();
-        });
-      });
+  ];
+
+  function run_tests(i, commandName, level) {
+    expect(info.commands.started).to.have.a.lengthOf(i + 1);
+    expect(info.commands.succeeded).to.have.a.lengthOf(i + 1);
+    expect(info.commands.started[i]).to.have.property('commandName', commandName);
+    expect(info.commands.succeeded[i]).to.have.property('commandName', commandName);
+    if (level !== undefined) {
+      expect(info.commands.started[i]).to.have.nested.property('command.readConcern.level', level);
+    } else {
+      expect(info.commands.started[i].command.readConcern).to.equal(undefined);
     }
-  });
-
-  it('Should set majority readConcern using MongoClient with options', {
-    metadata: { requires: { topology: 'replicaset', mongodb: '>= 3.2' } },
-
-    test: function(done) {
-      var configuration = this.configuration;
-      var listener = require('../..').instrument(function(err) {
-        test.equal(null, err);
-      });
-
-      // Contains all the apm events
-      var started = [];
-      var url = configuration.url();
-      var options = {
-        readConcern: {
-          level: 'majority'
+  }
+
+  tests.forEach(test => {
+    it(test.description, {
+      metadata: { requires: { topology: 'replicaset', mongodb: '>= 3.2' } },
+
+      test: function(done) {
+        // Get a new instance
+        const configuration = this.configuration;
+        client = !test.urlOptions
+          ? configuration.newClient({ w: 1 }, { poolSize: 1, readConcern: { level: test.level } })
+          : null;
+
+        if (test.urlReadConcernLevel || test.urlOptions) {
+          url = configuration.url();
+          if (!test.urlOptions) {
+            url =
+              url.indexOf('?') !== -1
+                ? f('%s&%s', url, test.urlReadConcernLevel)
+                : f('%s?%s', url, test.urlReadConcernLevel);
+          } else {
+            const options = {
+              readConcern: {
+                level: test.level
+              }
+            };
+            client = configuration.newClient(url, options);
```

Comment:
I think this section of code could be refactored so we only have one `client = configuration.newClient`. It looks like this is the same as line 98.